src/stdlib_math.fypp

```diff
@@ -1,21 +1,151 @@
+
+        ${t1}$ :: res(n)
+      end function ${RName}$
+    #:endfor
+
+
+    ! Add support for integer linspace
+    !! Version: Experimental
+    !!
+    !! Create rank 1 array of linearly spaced elements from start to end.
+    !! If the number of elements is not specified, create an array with size 100. If n is a negative value,
+    !! return an array with size 0. If n = 1, return end. When dealing with integers as the `start` and `end`
+    !! parameters, the return type is always a double precision real.
+    !!
+    !!([Specification](../page/specs/stdlib_math.html#description))
+    #:for k1, t1 in INT_KINDS_TYPES
+      #:set RName = rname("linspace_default", 1, t1, k1)
+      #! The interface for INT_KINDS_TYPES cannot be combined with RC_KINDS_TYPES
+      #! because the output for integer types is always a real with dp.
+      module function ${RName}$(start, end) result(res)
+        ${t1}$, intent(in) :: start
+        ${t1}$, intent(in) :: end
+
+        real(dp) :: res(DEFAULT_LINSPACE_LENGTH)
+      end function ${RName}$
+    #:endfor
+
+    #:for k1, t1 in INT_KINDS_TYPES
+      #:set RName = rname("linspace_n", 1, t1, k1)
+      module function ${RName}$(start, end, n) result(res)
+        ${t1}$, intent(in) :: start
+        ${t1}$, intent(in) :: end
+        integer, intent(in) :: n
+
+        real(dp) :: res(n)
+      end function ${RName}$
+    #:endfor
+
+  end interface
+
+  interface logspace
+  !! Version: Experimental
+  !!
+  !! Create rank 1 array of logarithmically spaced elements from base**start to base**end.
+  !! If the number of elements is not specified, create an array with size 50. If n is a negative value,
+  !! return an array with size 0. If n = 1, return base**end. If no base is specified, logspace will default to using a base of
+  !! 10
+  #:for k1, t1 in IRC_KINDS_TYPES
+    #:set RName = rname("logspace_default", 1, t1, k1)
+    module function ${RName}$(start, end) result(res)
+
+      ${t1}$, intent(in) :: start
+      ${t1}$, intent(in) :: end
+
+      ${t1}$ :: res(DEFAULT_LOGSPACE_LENGTH)
+
+    end function ${RName}$
+  #:endfor
+
+  #:for k1, t1 in IRC_KINDS_TYPES
+    #:set RName = rname("logspace_n", 1, t1, k1)
+    module function ${RName}$(start, end, n) result(res)
+      ${t1}$, intent(in) :: start
+      ${t1}$, intent(in) :: end
+      integer, intent(in) :: n
+
+      ${t1}$ :: res(n)
+    end function ${RName}$
+  #:endfor
+
+  #:for k1, t1 in IRC_KINDS_TYPES
+    #:for k2, t2 in IRC_KINDS_TYPES
+      #! k2, t2 correspond to the kind/type of the base that is passed
+      #:set RName = rname("logspace_n_base", 1, k1, k2)
+      #! ================================================================ !#
+      #! Need another function where base is not optional, otherwise the
+      #! compiler can not differentiate between
+      #! generic calls to logspace_n where a base is not present
+      #! ================================================================ !#
```

Comment:
I don't think we need this function, why can't we just add the base argument as optional one to the `logspace` function?

Also the current double generator loop seems a bit overkill too me, as it generates a lot of argument combinations that are unlikely to be used in practice (like single precision start end, with a quadruple precision base). A single generator loop should be sufficient for most use cases.

Reply:
So when I was originally developing I tried to have the base argument as optional, but as indicated in the fypp comment, the compiler is unable to resolve generic calls to `logspace_n` when the base is not present. I don't have a good example as proof because I was writing it a month ago, but I can try and write up a mock version to show you what I mean.

As for the overkill comment - I agree. I agree that it seems like overkill to have a double generator loop... _However_, putting myself in the position of the user, I would **love** to be able to call `linspace` and `logspace` functions just like I can in `MATLAB`. I want to be able to call `logspace(1, 10, 5, EULERS_NUMBER)` without worrying about changing the `start` and `end` values to coincide with the precision of whatever base I pass. I want to be able to call `logspace(1, 5, 10, 2)` without being confused why the compiler won't accept an integer `base`. Basically, the tradeoff for ease-of-use client side is a double generator loop. The whole purpose of using `fypp` is to allow for permissible generic programming, and it seems bizarre to me to not use it's full capabilities.

Reply:
fyi, I will mark the pull request as ready for review once I have fixed the bugs and it successfully compiles on Ubuntu. Until then, don't worry about approving the CI tests.

Reply:
The fypp generated code introduces some compilation overhead, in the past we already had cases were excessive preprocessor usage let to break downs of user machines due to huge memory requirements (16GB+).

The integer + real base case sounds like a reasonable combination for me. Still, I don't think we should mix different real precisions here.

Reply:
Ok, I agree with your last statement. This is just speculation on my part, but I imagine the crazy overhead comes into play when generating generic functions operating on multidimensional arrays. Since these two implementations are only dealing with rank 1 arrays, I can't imagine that any serious overhead is introduced. 

Nevertheless, I will see what I can do about reducing obsolete combinations. It will make the `.fypp` source code significantly less compact, but if it allows me to keep the integer + real base combinations then I'll be happy :)

Reply:
Hi,

I've come up with this proposed scheme to reduce obsolete combinations.

![20210711_141308.jpg](https://user-images.githubusercontent.com/40283612/125205846-45223f00-e252-11eb-917d-8a65c7c65cf1.jpg)

To keep the generated overhead reduced, real and complex kinds will only be combined based on the same precision. Additionally, it makes more sense to only provide functional support for the default integer type. Adding support for all integer types would indeed be more comprehensive, but I think we can ignore this edge case.

I will implement rhis scheme and resolve this conversation unless you have any objections.

Reply:
Looks good to me.